### PR TITLE
[AIRFLOW-3552] Add ImapAttachmetToS3Operator

### PR DESCRIPTION
**Pull request by feluelle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3552
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR adds an operator that transfers a mail attachment into an s3 bucket.
It also:
- adds tests

**NOTE:** This operator only transfers the latest attachment by name.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by feluelle** [comment]
PTAL: @feng-tao @kaxil @ashb 

Especially because of this:
> NOTE: This operator only transfers the latest attachment by name.

I am thinking of changing this so that you can transfer multiple attachments that for example match the regular expression you pass in but I don't know if this is really a good idea and if so how would be the best way to implement this.

Something like adding `imap_latest_only` (set to False) + passing `s3_bucket` so that the attachments just keep its names looks very strange to me. :/ ..but I don't know any other way of doing this.

File: airflow/contrib/operators/imap_to_s3_transfer_operator.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    :type imap_check_regex: bool
+    :param s3_overwrite: If set overwrites the s3 key if already exists.
+    :type s3_overwrite: bool
+    :param imap_conn_id: The reference to the connection details of the mail server.
+    :type imap_conn_id: str
+    :param s3_conn_id: The reference to the s3 connection details.
+    :type s3_conn_id: str
+    """
+
+    def __init__(self,
```
Comment (Fokko) [inline airflow/contrib/operators/imap_to_s3_transfer_operator.py]: Can you add `@apply_defaults`? https://github.com/apache/airflow/blob/master/airflow/operators/presto_check_operator.py#L60

File: airflow/contrib/operators/imap_to_s3_transfer_operator.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                 imap_attachment_name,
+                 s3_key,
+                 imap_mail_folder='INBOX',
+                 imap_check_regex=False,
+                 s3_overwrite=False,
+                 imap_conn_id='imap_default',
+                 s3_conn_id='s3_default',
```
Comment (Fokko) [inline airflow/contrib/operators/imap_to_s3_transfer_operator.py]: Can you change this into `aws_default`?

**Review by Fokko — CHANGES_REQUESTED** [review]
Thanks @feluelle 

Looking good, but I have some minor comments. PTAL

File: airflow/contrib/operators/imap_to_s3_transfer_operator.py (hunk 1 above)
Comment (feluelle) [reply airflow/contrib/operators/imap_to_s3_transfer_operator.py]: Thanks @Fokko did not notice I was missing that. I will add that :)

**Comment by feluelle** [comment]
@Fokko resolved and also added template fields.
Please take another look when you got the time. Thanks :)

**Comment by feluelle** [comment]
This PR kinda depends on #4475 . @Fokko may you also take a look there? :)

**Comment by OmerJog** [comment]
Is there a naming policy ?
This PR introduce `ImapToS3TransferOperator` in file `imap_to_s3_transfer_operator.py`

However there are two definition

Without `Operator`:
`OracleToOracleTransfer  `in file `oracle_to_oracle_transfer.py`
`OracleToAzureDataLakeTransfer `in file `oracle_to_azure_data_lake_transfer.py`

With `Opeartor`:
`GoogleCloudStorageToGoogleCloudStorageTransferOperator `in file `gcs_to_gcs_transfer_operator.py`
`S3ToGoogleCloudStorageTransferOperator  `in file `s3_to_gcs_transfer_operator.py`

It's very confusing.

**Comment by ashb** [comment]
Most things have with~out~ Operator - I think the Oracle ones you mention are the exception (but I haven't checked all of them)

(Edit: meant to say _with_ operator)

**Comment by OmerJog** [comment]
@ashb Actually most with Operator.
I'm not sure why the Transfer operators are treated differently at the end they are still Operators like the rest:
`PostgresToGoogleCloudStorageOperator`
`SFTPToS3Operator`
etc..

Actually there is a 3rd kind   `HiveToDynamoDBTransferOperator `in file  `hive_to_dynamodb.py`  (No Transfer nor Operator in the file name).

I think there should be specific rules for this in `CONTRIBUTING.md` and that will be enforced when code-reviewing. But this discussion is out of scope of this specific PR.

**Comment by feluelle** [comment]
@OmerJog Absolutly agree. I don't like how this is currently done, too. 
Maybe you can open a seperate PR for this. So we can continue discussing it there or even better send a mail to the mailing list - especially for **Naming Conventions**.

**Comment by OmerJog** [comment]
I'm not on mailing list that usually too spamy for me. Just thought it's worth mentioning

**Comment by feluelle** [comment]
Please do not merge this yet. I need to adapt this regarding to latest changes from #4475 .

@ashb Do you think we you should be able to specify the `not_found_mode` here? Or not for that kind of an operator? (default is `raise` then)

**Review by ashb — COMMENTED** [review]
From the class name I wouldn't have expeceted this to be about attacmhents, but just the message. Maybe `ImapAttachmentToS3TransferOperator` is a little bit long `ImapAttachmentToS3Operator`

If the sole purpose of this operator is to transfer attachments, then failing when there are none probably makes sense.

**Comment by feluelle** [comment]
Hey @ashb, I changed the class name. Any other comments about this feature? :)

**Review by Fokko — APPROVED** [review]

**Review by ashb — APPROVED** [review]
Not a comment on this PR, more a general observation:

I'd like to find a way to make this sort of XToYOperator more generic.

Something like:

```
CopyOperator(
    source=ImapAttachmentSomething(imap_conn_id...),
    dest=S3Fs(....)
)
```

**Very** half baked idea, but supporting arbitrary sources and dests would be nice.

**Comment by feluelle** [comment]
@ashb I understand your idea, but I don't think this can't be more generic. It won't get _simpler_.

What I have done in for our custom airflow plugin is adding kinda like an interface:
```
class BaseTransferOperator(BaseOperator):

  def __init__(self, source_conn_id, destination_conn_id, *args, **kwargs):
    super().__init__(*args, **kwargs)
```

**So all transfer operators need to have two connections.**

**Comment by mik-laj** [comment]
@ashb Look at:
https://github.com/apache/airflow/pull/3526
Here was the beginning of work on this idea.

**Comment by mik-laj** [comment]
@ashb 
We need to think about classified different types of system (ex. filesystem. database) and on the basis of this write operators, which will copy data between different types of systems.
Currently, we have `GenericTransfer` which copies between the database and the database.
Another operator can be created, which will copy data between database and file systems.
However, this requires a lot of deeper architectural thought.

**Pull request merged**